### PR TITLE
Use feature name in prediction

### PR DESCRIPTION
**Pull request by WWLoveBasketball**
fix issue #980 : Use feature name in the prediction

Todo: this PR only can deal the situation with label index is 0 (the first one).

### CODE REVIEW

File: src/application/predictor.hpp [hunk 1]
```diff
@@ -104,12 +108,45 @@ class Predictor {
...
+        for(int i = 0; i < static_cast<int>(header.size()); ++i) {
+          for(int j = 0; j < static_cast<int>(boosting_->FeatureNames().size()); ++j) {
+            if(header[i] == boosting_->FeatureNames()[j]) {
+              feature_names_map_[i] = j;
+              break;
+            }
+          }
+        }
+      }
+    }
```
Comment (guolinke) [inline src/application/predictor.hpp]: these is too heavy, refer to : https://github.com/Microsoft/LightGBM/blob/master/src/io/dataset_loader.cpp#L29-L32

File: src/application/predictor.hpp [hunk 2]
```diff
@@ -104,12 +108,45 @@ class Predictor {
...
+    parser_fun = [this, &parser, &tmp_label, &has_header, &feature_names_map_]
     (const char* buffer, std::vector<std::pair<int, double>>* feature) {
       parser->ParseOneLine(buffer, feature, &tmp_label);
+      if(has_header) {
+        for(int i = 0; i < static_cast<int>(feature->size()); ++i) {
+          if(feature_names_map_.find((*feature)[i].first) != feature_names_map_.end()) {
+            (*feature)[i].first = feature_names_map_[(*feature)[i].first];
+          }
+          else {
+            (*feature).erase((*feature).begin() + i);
```
Comment (guolinke) [inline src/application/predictor.hpp]: erase in the loop ?

Comment (WWLoveBasketball) [reply src/application/predictor.hpp]: yes, here is for deleting the features which are not used in the model but exist in the test data.

Comment (guolinke) [reply src/application/predictor.hpp]: I don't think it is safe to do this. 
A better solution is allocating another map that used to store non-deleted items.

Comment (WWLoveBasketball) [reply src/application/predictor.hpp]: I have moved the non-used features to the end of the feature vector, I think this way is also a safe solution.

File: src/application/predictor.hpp [hunk 3]
```diff
@@ -104,12 +105,48 @@ class Predictor {
...
       parser->ParseOneLine(buffer, feature, &tmp_label);
+      if(need_adjust) {
+        int j = 0;
+        for(int i = 0; i < static_cast<int>(feature_names_map_.size()); ++i) {
+          if(feature_names_map_.find((*feature)[i].first) != feature_names_map_.end()) {
+            (*feature)[i].first = feature_names_map_[(*feature)[i].first];
+          }
+          else {
+            //move the non-used features to the end of the feature vector
+            std::swap((*feature)[i], (*feature)[static_cast<int>(feature_names_map_.size())+(j++)]);
```
Comment (guolinke) [inline src/application/predictor.hpp]: I think ```feature``` is the sparse pair (only store the non-zero features). So its length may is not equal to ```feature_names_map_``` .

**Pull request merged**